{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Objective: To organize hydrophobicity scales\n",
    "\n",
    "In this notebook, I am simply combining hydrophobicity scales from different sources. \n",
    "The first source was obtained from the Qiagen website [here](https://resources.qiagenbioinformatics.com/manuals/clcgenomicsworkbench/650/Hydrophobicity_scales.html).\n",
    "These are directly from the literature. \n",
    "\n",
    "The second source is from a paper *Comparison of hydrophobicity scales for predicting biophysical properties of antibodies* by Waibl et al. Frontiers in Molecular Biosciences 2022. In this paper, they take the scales from the literature and scale them by adding a constant such that Glycine hydrophobicity is 0 and then scale such that the variance is 1. (ie. $z_{aa}=\\frac{x_{aa}-x_{\\text{gly}}}{\\mathbb{V}(X)}$), where $x_aa$ is the hydrophobicity score for a given amino acid. \n",
    "\n",
    "I took the table directly from the paper as text. Below, I clean it up and join it to the first source. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from io import StringIO\n",
    "from pathlib import Path\n",
    "import developability"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Code for cleaning the text "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_df_from_num_list(nums, header):\n",
    "    num_list = [header]\n",
    "    current_list = []\n",
    "    i=0\n",
    "    for num in nums.split(','):\n",
    "        \n",
    "        i+=1\n",
    "        current_list.append(num)\n",
    "        if i//3:\n",
    "            num_list.append(','.join(current_list))\n",
    "            current_list = []\n",
    "            i=0\n",
    "    num_list = '\\n'.join(num_list)\n",
    "    return pd.read_csv(StringIO(num_list), sep=\",\", index_col=False)\n",
    "\n",
    "def clean_sign(df):\n",
    "    for col in df.columns:\n",
    "        vals = []\n",
    "        for val in df[col].values:\n",
    "            if val[0] == '−':\n",
    "                vals.append(float(val[1:])*-1)\n",
    "            else:\n",
    "                vals.append(float(val))\n",
    "        df[col] = vals\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "text =\"\"\"Residue,BaMe,BlMo,Ei\n",
    "ALA,0.75,0.37,0.15,\n",
    "ARG,−0.02,−1.52,−3.09,\n",
    "ASN,−0.16,−0.79,−1.29,\n",
    "ASP,−0.50,−1.43,−1.42,\n",
    "CYS,2.60,0.55,−0.19,\n",
    "GLN,−0.11,−0.76,−1.37,\n",
    "GLU,−0.54,−1.40,−1.26,\n",
    "GLY,0.00,0.00,0.00,\n",
    "HIS,0.57,−1.00,−0.90,\n",
    "ILE,2.19,1.34,0.92,\n",
    "LEU,1.97,1.34,0.60,\n",
    "LYS,−0.90,−0.67,−2.03,\n",
    "MET,1.22,0.73,0.16,\n",
    "PHE,1.92,1.52,0.73,\n",
    "PRO,0.72,0.64,−0.37,\n",
    "SER,0.11,−0.43,−0.68,\n",
    "THR,0.47,−0.15,−0.55,\n",
    "TRP,1.51,1.16,0.34,\n",
    "TYR,1.36,1.16,−0.23,\n",
    "VAL,1.88,1.00,0.61\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(StringIO(text), sep=\",\", index_col=False) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "header1 = \"\"\"KyDo,Me,Ro\"\"\"\n",
    "nums1=\"\"\"0.76,0.07,0.18,−1.41,0.11,−0.71,−1.06,0.11,−0.80,−1.06,−1.08,−0.89,1.00,−0.90,1.69,−1.06,−0.63,−0.89,−1.06,−2.23,−0.89,0.00,0.00,0.00,−0.96,−0.46,0.53,1.68,1.83,1.42,1.44,1.16,1.16,−1.20,0.01,−1.78,0.79,0.63,1.16,1.10,1.74,1.42,−0.41,0.80,−0.71,−0.14,0.16,−0.53,−0.10,0.36,−0.18,−0.17,1.96,1.16,−0.31,0.80,0.36,1.58,0.36,1.25\"\"\"\n",
    "df2 = create_df_from_num_list(nums1, header1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "header2 = \"WiWh,Ja,Mi\"\n",
    "nums2 = \"−0.20,0.06,0.40,−0.41,−0.32,−0.15,−0.51,0.13,−0.37,−1.53,−0.43,−0.43,0.31,0.55,1.65,−0.71,0.46,−0.29,−2.51,−0.72,−0.40,0.00,0.00,0.00,−0.20,0.03,0.30,0.35,1.54,2.08,0.71,1.54,1.91,−0.59,−1.07,−0.74,0.30,0.49,2.14,1.43,2.48,2.18,−0.55,0.44,−0.29,−0.15,0.07,−0.19,−0.16,0.16,0.00,2.33,2.81,1.52,1.19,1.84,0.68,−0.08,0.97,1.51\"\n",
    "df3 = create_df_from_num_list(nums2, header2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "df4 = (pd.concat([df, df2, df3], axis=1)\n",
    "       .set_index('Residue')\n",
    "       )\n",
    "df4 = clean_sign(df4)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "df4.to_csv('output.csv')\n",
    "df5 = pd.read_csv('output.csv', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "BaMe    float64\n",
       "BlMo    float64\n",
       "Ei      float64\n",
       "KyDo    float64\n",
       "Me      float64\n",
       "Ro      float64\n",
       "WiWh    float64\n",
       "Ja      float64\n",
       "Mi      float64\n",
       "dtype: object"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df5.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>amino_acid_name</th>\n",
       "      <th>Kyte_Doolittle</th>\n",
       "      <th>Hopp_Woods</th>\n",
       "      <th>Cornette</th>\n",
       "      <th>Eisenberg</th>\n",
       "      <th>Rose</th>\n",
       "      <th>Janin</th>\n",
       "      <th>Engelman_GES</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aa</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>Alanine</td>\n",
       "      <td>1.8</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.62</td>\n",
       "      <td>0.74</td>\n",
       "      <td>0.3</td>\n",
       "      <td>1.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R</th>\n",
       "      <td>Arginine</td>\n",
       "      <td>-4.5</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.4</td>\n",
       "      <td>-2.53</td>\n",
       "      <td>0.64</td>\n",
       "      <td>-1.4</td>\n",
       "      <td>-12.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <td>Asparagine</td>\n",
       "      <td>-3.5</td>\n",
       "      <td>0.2</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>-0.78</td>\n",
       "      <td>0.63</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>-4.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>Aspartic acid</td>\n",
       "      <td>-3.5</td>\n",
       "      <td>3.0</td>\n",
       "      <td>-3.1</td>\n",
       "      <td>-0.90</td>\n",
       "      <td>0.62</td>\n",
       "      <td>-0.6</td>\n",
       "      <td>-9.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>Cysteine</td>\n",
       "      <td>2.5</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>4.1</td>\n",
       "      <td>0.29</td>\n",
       "      <td>0.91</td>\n",
       "      <td>0.9</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   amino_acid_name  Kyte_Doolittle  Hopp_Woods  Cornette  Eisenberg  Rose  \\\n",
       "aa                                                                          \n",
       "A          Alanine             1.8        -0.5       0.2       0.62  0.74   \n",
       "R         Arginine            -4.5         3.0       1.4      -2.53  0.64   \n",
       "N       Asparagine            -3.5         0.2      -0.5      -0.78  0.63   \n",
       "D    Aspartic acid            -3.5         3.0      -3.1      -0.90  0.62   \n",
       "C         Cysteine             2.5        -1.0       4.1       0.29  0.91   \n",
       "\n",
       "    Janin  Engelman_GES  \n",
       "aa                       \n",
       "A     0.3           1.6  \n",
       "R    -1.4         -12.3  \n",
       "N    -0.5          -4.8  \n",
       "D    -0.6          -9.2  \n",
       "C     0.9           2.0  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_path = Path(developability.__path__[0])\n",
    "scales = pd.read_csv(data_path/'hydrophobicity_scales.csv', index_col=0).sort_values('amino_acid_name')\n",
    "scales.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Check scaling method from paper\n",
    "In the paper mentioned above, they scaled the hydrophobicity per residue scales by adding a constant to center the values by gylcine with glycine set to zero. They then standardized the variance. Below, I check that. \n",
    "\n",
    "Although not perfect, it seems my understanding of their method is roughly correct. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([-0.01,  0.08,  0.03,  0.04, -0.  ,  0.15, -0.07,  0.  ,  0.02,\n",
       "       -0.02, -0.02,  0.05,  0.  , -0.02,  0.01,  0.02,  0.02, -0.01,\n",
       "        0.01, -0.01])"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ei = scales['Eisenberg']\n",
    "ei2 = (ei- ei.loc['G'])/(ei- ei.loc['G']).std()\n",
    "Ei = df5['Ei']\n",
    "\n",
    "np.round((ei2.values - Ei.values),2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([-0.02,  0.04,  0.02,  0.02, -0.03,  0.02,  0.02,  0.  ,  0.02,\n",
       "       -0.04, -0.03,  0.03, -0.02, -0.03,  0.01,  0.01, -0.  ,  0.  ,\n",
       "        0.01, -0.04])"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kd = scales['Kyte_Doolittle']\n",
    "kd2 = (kd- kd.loc['G'])/(kd- kd.loc['G']).std()\n",
    "Kd = df5['KyDo']\n",
    "\n",
    "np.round(kd2.values - Kd.values,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>amino_acid_name</th>\n",
       "      <th>Residue</th>\n",
       "      <th>Kyte_Doolittle</th>\n",
       "      <th>Hopp_Woods</th>\n",
       "      <th>Cornette</th>\n",
       "      <th>Eisenberg</th>\n",
       "      <th>Rose</th>\n",
       "      <th>Janin</th>\n",
       "      <th>Engelman_GES</th>\n",
       "      <th>BaMe</th>\n",
       "      <th>BlMo</th>\n",
       "      <th>Ei</th>\n",
       "      <th>KyDo</th>\n",
       "      <th>Me</th>\n",
       "      <th>Ro</th>\n",
       "      <th>WiWh</th>\n",
       "      <th>Ja</th>\n",
       "      <th>Mi</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aa</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>Alanine</td>\n",
       "      <td>ALA</td>\n",
       "      <td>1.8</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.62</td>\n",
       "      <td>0.74</td>\n",
       "      <td>0.3</td>\n",
       "      <td>1.6</td>\n",
       "      <td>0.75</td>\n",
       "      <td>0.37</td>\n",
       "      <td>0.15</td>\n",
       "      <td>0.76</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.18</td>\n",
       "      <td>-0.20</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R</th>\n",
       "      <td>Arginine</td>\n",
       "      <td>ARG</td>\n",
       "      <td>-4.5</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.4</td>\n",
       "      <td>-2.53</td>\n",
       "      <td>0.64</td>\n",
       "      <td>-1.4</td>\n",
       "      <td>-12.3</td>\n",
       "      <td>-0.02</td>\n",
       "      <td>-1.52</td>\n",
       "      <td>-3.09</td>\n",
       "      <td>-1.41</td>\n",
       "      <td>0.11</td>\n",
       "      <td>-0.71</td>\n",
       "      <td>-0.41</td>\n",
       "      <td>-0.32</td>\n",
       "      <td>-0.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <td>Asparagine</td>\n",
       "      <td>ASN</td>\n",
       "      <td>-3.5</td>\n",
       "      <td>0.2</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>-0.78</td>\n",
       "      <td>0.63</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>-4.8</td>\n",
       "      <td>-0.16</td>\n",
       "      <td>-0.79</td>\n",
       "      <td>-1.29</td>\n",
       "      <td>-1.06</td>\n",
       "      <td>0.11</td>\n",
       "      <td>-0.80</td>\n",
       "      <td>-0.51</td>\n",
       "      <td>0.13</td>\n",
       "      <td>-0.37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>Aspartic acid</td>\n",
       "      <td>ASP</td>\n",
       "      <td>-3.5</td>\n",
       "      <td>3.0</td>\n",
       "      <td>-3.1</td>\n",
       "      <td>-0.90</td>\n",
       "      <td>0.62</td>\n",
       "      <td>-0.6</td>\n",
       "      <td>-9.2</td>\n",
       "      <td>-0.50</td>\n",
       "      <td>-1.43</td>\n",
       "      <td>-1.42</td>\n",
       "      <td>-1.06</td>\n",
       "      <td>-1.08</td>\n",
       "      <td>-0.89</td>\n",
       "      <td>-1.53</td>\n",
       "      <td>-0.43</td>\n",
       "      <td>-0.43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>Cysteine</td>\n",
       "      <td>CYS</td>\n",
       "      <td>2.5</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>4.1</td>\n",
       "      <td>0.29</td>\n",
       "      <td>0.91</td>\n",
       "      <td>0.9</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.60</td>\n",
       "      <td>0.55</td>\n",
       "      <td>-0.19</td>\n",
       "      <td>1.00</td>\n",
       "      <td>-0.90</td>\n",
       "      <td>1.69</td>\n",
       "      <td>0.31</td>\n",
       "      <td>0.55</td>\n",
       "      <td>1.65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>Glutamic acid</td>\n",
       "      <td>GLN</td>\n",
       "      <td>-3.5</td>\n",
       "      <td>3.0</td>\n",
       "      <td>-1.8</td>\n",
       "      <td>-0.74</td>\n",
       "      <td>0.62</td>\n",
       "      <td>-0.7</td>\n",
       "      <td>-8.2</td>\n",
       "      <td>-0.11</td>\n",
       "      <td>-0.76</td>\n",
       "      <td>-1.37</td>\n",
       "      <td>-1.06</td>\n",
       "      <td>-0.63</td>\n",
       "      <td>-0.89</td>\n",
       "      <td>-0.71</td>\n",
       "      <td>0.46</td>\n",
       "      <td>-0.29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q</th>\n",
       "      <td>Glutamine</td>\n",
       "      <td>GLU</td>\n",
       "      <td>-3.5</td>\n",
       "      <td>0.2</td>\n",
       "      <td>-2.8</td>\n",
       "      <td>-0.85</td>\n",
       "      <td>0.62</td>\n",
       "      <td>-0.7</td>\n",
       "      <td>-4.1</td>\n",
       "      <td>-0.54</td>\n",
       "      <td>-1.40</td>\n",
       "      <td>-1.26</td>\n",
       "      <td>-1.06</td>\n",
       "      <td>-2.23</td>\n",
       "      <td>-0.89</td>\n",
       "      <td>-2.51</td>\n",
       "      <td>-0.72</td>\n",
       "      <td>-0.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>Glycine</td>\n",
       "      <td>GLY</td>\n",
       "      <td>-0.4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.72</td>\n",
       "      <td>0.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H</th>\n",
       "      <td>Histidine</td>\n",
       "      <td>HIS</td>\n",
       "      <td>-3.2</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>-0.40</td>\n",
       "      <td>0.78</td>\n",
       "      <td>-0.1</td>\n",
       "      <td>-3.0</td>\n",
       "      <td>0.57</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>-0.90</td>\n",
       "      <td>-0.96</td>\n",
       "      <td>-0.46</td>\n",
       "      <td>0.53</td>\n",
       "      <td>-0.20</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I</th>\n",
       "      <td>Isoleucine</td>\n",
       "      <td>ILE</td>\n",
       "      <td>4.5</td>\n",
       "      <td>-1.8</td>\n",
       "      <td>4.8</td>\n",
       "      <td>1.38</td>\n",
       "      <td>0.88</td>\n",
       "      <td>0.7</td>\n",
       "      <td>3.1</td>\n",
       "      <td>2.19</td>\n",
       "      <td>1.34</td>\n",
       "      <td>0.92</td>\n",
       "      <td>1.68</td>\n",
       "      <td>1.83</td>\n",
       "      <td>1.42</td>\n",
       "      <td>0.35</td>\n",
       "      <td>1.54</td>\n",
       "      <td>2.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L</th>\n",
       "      <td>Leucine</td>\n",
       "      <td>LEU</td>\n",
       "      <td>3.8</td>\n",
       "      <td>-1.8</td>\n",
       "      <td>5.7</td>\n",
       "      <td>1.06</td>\n",
       "      <td>0.85</td>\n",
       "      <td>0.5</td>\n",
       "      <td>2.8</td>\n",
       "      <td>1.97</td>\n",
       "      <td>1.34</td>\n",
       "      <td>0.60</td>\n",
       "      <td>1.44</td>\n",
       "      <td>1.16</td>\n",
       "      <td>1.16</td>\n",
       "      <td>0.71</td>\n",
       "      <td>1.54</td>\n",
       "      <td>1.91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K</th>\n",
       "      <td>Lysine</td>\n",
       "      <td>LYS</td>\n",
       "      <td>-3.9</td>\n",
       "      <td>3.0</td>\n",
       "      <td>-3.1</td>\n",
       "      <td>-1.50</td>\n",
       "      <td>0.52</td>\n",
       "      <td>-1.8</td>\n",
       "      <td>-8.8</td>\n",
       "      <td>-0.90</td>\n",
       "      <td>-0.67</td>\n",
       "      <td>-2.03</td>\n",
       "      <td>-1.20</td>\n",
       "      <td>0.01</td>\n",
       "      <td>-1.78</td>\n",
       "      <td>-0.59</td>\n",
       "      <td>-1.07</td>\n",
       "      <td>-0.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>Methionine</td>\n",
       "      <td>MET</td>\n",
       "      <td>1.9</td>\n",
       "      <td>-1.3</td>\n",
       "      <td>4.2</td>\n",
       "      <td>0.64</td>\n",
       "      <td>0.85</td>\n",
       "      <td>0.4</td>\n",
       "      <td>3.4</td>\n",
       "      <td>1.22</td>\n",
       "      <td>0.73</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.79</td>\n",
       "      <td>0.63</td>\n",
       "      <td>1.16</td>\n",
       "      <td>0.30</td>\n",
       "      <td>0.49</td>\n",
       "      <td>2.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F</th>\n",
       "      <td>Phenylalanine</td>\n",
       "      <td>PHE</td>\n",
       "      <td>2.8</td>\n",
       "      <td>-2.5</td>\n",
       "      <td>4.4</td>\n",
       "      <td>1.19</td>\n",
       "      <td>0.88</td>\n",
       "      <td>0.5</td>\n",
       "      <td>3.7</td>\n",
       "      <td>1.92</td>\n",
       "      <td>1.52</td>\n",
       "      <td>0.73</td>\n",
       "      <td>1.10</td>\n",
       "      <td>1.74</td>\n",
       "      <td>1.42</td>\n",
       "      <td>1.43</td>\n",
       "      <td>2.48</td>\n",
       "      <td>2.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P</th>\n",
       "      <td>Proline</td>\n",
       "      <td>PRO</td>\n",
       "      <td>-1.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-2.2</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.64</td>\n",
       "      <td>-0.3</td>\n",
       "      <td>-0.2</td>\n",
       "      <td>0.72</td>\n",
       "      <td>0.64</td>\n",
       "      <td>-0.37</td>\n",
       "      <td>-0.41</td>\n",
       "      <td>0.80</td>\n",
       "      <td>-0.71</td>\n",
       "      <td>-0.55</td>\n",
       "      <td>0.44</td>\n",
       "      <td>-0.29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>Serine</td>\n",
       "      <td>SER</td>\n",
       "      <td>-0.8</td>\n",
       "      <td>0.3</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>-0.18</td>\n",
       "      <td>0.66</td>\n",
       "      <td>-0.1</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.11</td>\n",
       "      <td>-0.43</td>\n",
       "      <td>-0.68</td>\n",
       "      <td>-0.14</td>\n",
       "      <td>0.16</td>\n",
       "      <td>-0.53</td>\n",
       "      <td>-0.15</td>\n",
       "      <td>0.07</td>\n",
       "      <td>-0.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T</th>\n",
       "      <td>Threonine</td>\n",
       "      <td>THR</td>\n",
       "      <td>-0.7</td>\n",
       "      <td>-0.4</td>\n",
       "      <td>-1.9</td>\n",
       "      <td>-0.05</td>\n",
       "      <td>0.70</td>\n",
       "      <td>-0.2</td>\n",
       "      <td>1.2</td>\n",
       "      <td>0.47</td>\n",
       "      <td>-0.15</td>\n",
       "      <td>-0.55</td>\n",
       "      <td>-0.10</td>\n",
       "      <td>0.36</td>\n",
       "      <td>-0.18</td>\n",
       "      <td>-0.16</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>W</th>\n",
       "      <td>Tryptophan</td>\n",
       "      <td>TRP</td>\n",
       "      <td>-0.9</td>\n",
       "      <td>-3.4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.81</td>\n",
       "      <td>0.85</td>\n",
       "      <td>0.3</td>\n",
       "      <td>1.9</td>\n",
       "      <td>1.51</td>\n",
       "      <td>1.16</td>\n",
       "      <td>0.34</td>\n",
       "      <td>-0.17</td>\n",
       "      <td>1.96</td>\n",
       "      <td>1.16</td>\n",
       "      <td>2.33</td>\n",
       "      <td>2.81</td>\n",
       "      <td>1.52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y</th>\n",
       "      <td>Tyrosine</td>\n",
       "      <td>TYR</td>\n",
       "      <td>-1.3</td>\n",
       "      <td>-2.3</td>\n",
       "      <td>3.2</td>\n",
       "      <td>0.26</td>\n",
       "      <td>0.76</td>\n",
       "      <td>-0.4</td>\n",
       "      <td>-0.7</td>\n",
       "      <td>1.36</td>\n",
       "      <td>1.16</td>\n",
       "      <td>-0.23</td>\n",
       "      <td>-0.31</td>\n",
       "      <td>0.80</td>\n",
       "      <td>0.36</td>\n",
       "      <td>1.19</td>\n",
       "      <td>1.84</td>\n",
       "      <td>0.68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V</th>\n",
       "      <td>Valine</td>\n",
       "      <td>VAL</td>\n",
       "      <td>4.2</td>\n",
       "      <td>-1.5</td>\n",
       "      <td>4.7</td>\n",
       "      <td>1.08</td>\n",
       "      <td>0.86</td>\n",
       "      <td>0.6</td>\n",
       "      <td>2.6</td>\n",
       "      <td>1.88</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.61</td>\n",
       "      <td>1.58</td>\n",
       "      <td>0.36</td>\n",
       "      <td>1.25</td>\n",
       "      <td>-0.08</td>\n",
       "      <td>0.97</td>\n",
       "      <td>1.51</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   amino_acid_name Residue  Kyte_Doolittle  Hopp_Woods  Cornette  Eisenberg  \\\n",
       "aa                                                                            \n",
       "A          Alanine     ALA             1.8        -0.5       0.2       0.62   \n",
       "R         Arginine     ARG            -4.5         3.0       1.4      -2.53   \n",
       "N       Asparagine     ASN            -3.5         0.2      -0.5      -0.78   \n",
       "D    Aspartic acid     ASP            -3.5         3.0      -3.1      -0.90   \n",
       "C         Cysteine     CYS             2.5        -1.0       4.1       0.29   \n",
       "E    Glutamic acid     GLN            -3.5         3.0      -1.8      -0.74   \n",
       "Q        Glutamine     GLU            -3.5         0.2      -2.8      -0.85   \n",
       "G          Glycine     GLY            -0.4         0.0       0.0       0.48   \n",
       "H        Histidine     HIS            -3.2        -0.5       0.5      -0.40   \n",
       "I       Isoleucine     ILE             4.5        -1.8       4.8       1.38   \n",
       "L          Leucine     LEU             3.8        -1.8       5.7       1.06   \n",
       "K           Lysine     LYS            -3.9         3.0      -3.1      -1.50   \n",
       "M       Methionine     MET             1.9        -1.3       4.2       0.64   \n",
       "F    Phenylalanine     PHE             2.8        -2.5       4.4       1.19   \n",
       "P          Proline     PRO            -1.6         0.0      -2.2       0.12   \n",
       "S           Serine     SER            -0.8         0.3      -0.5      -0.18   \n",
       "T        Threonine     THR            -0.7        -0.4      -1.9      -0.05   \n",
       "W       Tryptophan     TRP            -0.9        -3.4       1.0       0.81   \n",
       "Y         Tyrosine     TYR            -1.3        -2.3       3.2       0.26   \n",
       "V           Valine     VAL             4.2        -1.5       4.7       1.08   \n",
       "\n",
       "    Rose  Janin  Engelman_GES  BaMe  BlMo    Ei  KyDo    Me    Ro  WiWh    Ja  \\\n",
       "aa                                                                              \n",
       "A   0.74    0.3           1.6  0.75  0.37  0.15  0.76  0.07  0.18 -0.20  0.06   \n",
       "R   0.64   -1.4         -12.3 -0.02 -1.52 -3.09 -1.41  0.11 -0.71 -0.41 -0.32   \n",
       "N   0.63   -0.5          -4.8 -0.16 -0.79 -1.29 -1.06  0.11 -0.80 -0.51  0.13   \n",
       "D   0.62   -0.6          -9.2 -0.50 -1.43 -1.42 -1.06 -1.08 -0.89 -1.53 -0.43   \n",
       "C   0.91    0.9           2.0  2.60  0.55 -0.19  1.00 -0.90  1.69  0.31  0.55   \n",
       "E   0.62   -0.7          -8.2 -0.11 -0.76 -1.37 -1.06 -0.63 -0.89 -0.71  0.46   \n",
       "Q   0.62   -0.7          -4.1 -0.54 -1.40 -1.26 -1.06 -2.23 -0.89 -2.51 -0.72   \n",
       "G   0.72    0.3           1.0  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00   \n",
       "H   0.78   -0.1          -3.0  0.57 -1.00 -0.90 -0.96 -0.46  0.53 -0.20  0.03   \n",
       "I   0.88    0.7           3.1  2.19  1.34  0.92  1.68  1.83  1.42  0.35  1.54   \n",
       "L   0.85    0.5           2.8  1.97  1.34  0.60  1.44  1.16  1.16  0.71  1.54   \n",
       "K   0.52   -1.8          -8.8 -0.90 -0.67 -2.03 -1.20  0.01 -1.78 -0.59 -1.07   \n",
       "M   0.85    0.4           3.4  1.22  0.73  0.16  0.79  0.63  1.16  0.30  0.49   \n",
       "F   0.88    0.5           3.7  1.92  1.52  0.73  1.10  1.74  1.42  1.43  2.48   \n",
       "P   0.64   -0.3          -0.2  0.72  0.64 -0.37 -0.41  0.80 -0.71 -0.55  0.44   \n",
       "S   0.66   -0.1           0.6  0.11 -0.43 -0.68 -0.14  0.16 -0.53 -0.15  0.07   \n",
       "T   0.70   -0.2           1.2  0.47 -0.15 -0.55 -0.10  0.36 -0.18 -0.16  0.16   \n",
       "W   0.85    0.3           1.9  1.51  1.16  0.34 -0.17  1.96  1.16  2.33  2.81   \n",
       "Y   0.76   -0.4          -0.7  1.36  1.16 -0.23 -0.31  0.80  0.36  1.19  1.84   \n",
       "V   0.86    0.6           2.6  1.88  1.00  0.61  1.58  0.36  1.25 -0.08  0.97   \n",
       "\n",
       "      Mi  \n",
       "aa        \n",
       "A   0.40  \n",
       "R  -0.15  \n",
       "N  -0.37  \n",
       "D  -0.43  \n",
       "C   1.65  \n",
       "E  -0.29  \n",
       "Q  -0.40  \n",
       "G   0.00  \n",
       "H   0.30  \n",
       "I   2.08  \n",
       "L   1.91  \n",
       "K  -0.74  \n",
       "M   2.14  \n",
       "F   2.18  \n",
       "P  -0.29  \n",
       "S  -0.19  \n",
       "T   0.00  \n",
       "W   1.52  \n",
       "Y   0.68  \n",
       "V   1.51  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scales2 = pd.concat([scales.reset_index(), df5.reset_index()], axis=1)\n",
    "order = ['aa', 'amino_acid_name', 'Residue']\n",
    "order = order + [col for col in scales2.columns if col not in order]\n",
    "scales2 = scales2[order].set_index('aa')\n",
    "scales2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Save them to csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "scales.to_csv(data_path/'hydrophobicity_scales_old.csv')\n",
    "scales2.to_csv(data_path/'hydrophobicity_scales.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
